### ISSUE TITLE
APEv2 tags created by MP3Gain Express

### ISSUE THREAD

**Issue by jri** [Borewit/music-metadata#257]
**Feature Request**
I want read the ReplayGain tags created by [MP3Gain Express](http://projects.sappharad.com/mp3gain/) (a macOS GUI for `mp3gain`) in MP3 files. While `mp3gain` provides an option to create ID3v2 tags instead of APEv2 tags (the default), MP3Gain Express does not. So my entire collection is MP3 with APEv2 tags.

All I need is a simple command line tool to read out the MP3's APEv2 tags. I've put some effort in finding such a tool but without success. I'm excited about the quality of `music-metadata` and the effort you put in it. I would love to use it.

This is a follow-up of #191

**Comment by Borewit** [comment]
The way you describe it, it seems you want to go through your files ones. 

Can you get the job done with the branch [hack-reading-mp3-with-ape](https://github.com/Borewit/music-metadata/tree/hack-reading-mp3-with-ape)?

**Comment by jri** [comment]
> The way you describe it, it seems you want to go through your files ones.

Yes, I don't need the stream capability, just read files.

Indeed with that branch reading APE tags in MP3s kind of work! But not fully.
I've tested 3 files:

With `04 Aperitif de la mort.mp3` it works.
`tagTypes` properly lists [ 'ID3v2.2', 'APEv2' ] and native tags are returned:
```
APEv2:
      [ { id: 'MP3GAIN_MINMAX', value: '108,181' },
        { id: 'MP3GAIN_UNDO', value: '+003,+003,N' },
        { id: 'REPLAYGAIN_TRACK_GAIN', value: '+0.205000 dB' },
        { id: 'REPLAYGAIN_TRACK_PEAK', value: '0.604713' } ]
```

That's great!

However for 2nd file `open relationship 01 referents.mp3` tagTypes only lists [ 'ID3v2.3' ] despite it definitely has APE tags. (`mp3val` lists `+ID3v1+ID3v2+APEv2`)

And for 3rd file `15 - Oh Madonna (You Stole My Pants).mp3` an error is thrown: `AssertionError [ERR_ASSERTION]: APEv2 Footer preamble`.

For testing I'm using your example program:
```js
const mm = require('music-metadata')
const util = require('util')

mm.parseFile('APE-MP3-TEST/15 - Oh Madonna (You Stole My Pants).mp3', {native: true})
  .then(metadata => {
    console.log(util.inspect(metadata, {showHidden: false, depth: null}))
  })
  .catch(err => {
    console.error(err)
  });
```

Here are the 3 test files: 
[APE-MP3-TEST.zip](https://github.com/Borewit/music-metadata/files/3612202/APE-MP3-TEST.zip)

Thank you very much for quick action!

**Comment by Borewit** [comment]
A combination ID3v1 and APEv2 headers, I expected that one to come.

**Comment by Borewit** [comment]
Can you try again @jri?

**Comment by jri** [comment]
Indeed the 2nd test file now works!
tagTypes properly lists [ 'ID3v2.3', 'APEv2', 'ID3v1' ] and APE tags are included:
```
APEv2:
      [ { id: 'MP3GAIN_MINMAX', value: '136,248' },
        { id: 'MP3GAIN_UNDO', value: '+007,+007,N' },
        { id: 'REPLAYGAIN_TRACK_GAIN', value: '-0.305000 dB' },
        { id: 'REPLAYGAIN_TRACK_PEAK', value: '0.294162' } ]
```

Also the 3rd test file (which throwed an error before) now works!
tagTypes are also [ 'ID3v2.3', 'APEv2', 'ID3v1' ].
```
APEv2:
      [ { id: 'MP3GAIN_MINMAX', value: '111,184' },
        { id: 'REPLAYGAIN_TRACK_GAIN', value: '-0.440000 dB' },
        { id: 'REPLAYGAIN_TRACK_PEAK', value: '0.666330' } ]
```
In contrast to the other 2 files `MP3GAIN_UNDO` tag is missing here, but that's not a problem of music-metadata I guess.

(The 1st test file stills works as well.)

So it seems "mission accomplished".
That's great!

The quickest successful action I have experienced ever!
Thank you very much for your effort!

I'll let you know when I encounter further problematic files.

**Comment by Borewit** [comment]
There is indeed no `MP3GAIN_UNDO` tag in the "3rd" file:
![image](https://user-images.githubusercontent.com/23255260/64907396-1dafc900-d6f2-11e9-8699-b54f26f0799a.png)

I recommend to find a way normalizing your library on MP3 ID3v2.4 and get rid of all the other tag headers. Alternatively replace MP3 entirely with FLACs (probably needless to say, but please never ever convert MP3 or any other lossy format to FLAC).

Why are these _gain_, _peak_, _undo_ and _minmax_ tags important for you?

**Comment by jri** [comment]
I'm about to set up an icecast stream. One crucial requirement is pleasant perceived volume for every track. Over the course of several years I've applied per-track volume adjustments manually to the entire source material (mainly MP3 and AAC) by the means of iTunes. iTunes stores the manual adjustments mainly (but not only) as ID3v2 RVAD tags. (Over the years iTunes changed its storage approach several times, so the entire collection is not uniform). After some recent experiments with icecast and several tools I got the impression generating an icecast stream that respects my manual volume adjustments will become quite a challenge.

Now I've decided to give ReplayGain a try and have applied it to the entire source material (by the means of mp3gain via MP3Gain Express GUI). Now I want examine if mp3gain did a good job. Besides the actual listening experience I also want compare the applied gain values with my manual adjustments. Just for curiosity. (To me it's quite a step to dismiss the manual adjustments I've meticulously applied over the course of many years, in favor of an automatic solution.)

To my understanding I can tell the gain applied by mp3gain through the `MP3GAIN_UNDO` tag. When it says e.g. `'+004,+004,N'` I know the applied gain was -4 dB (or is it 6dB? I've read something about a factor of 1.5). Now music-metadata allows me to read that information and I'm happy :-)

Normalizing all tags to ID3v2.4 and/or replacing all MP3/AAC by FLAC is not an option at the moment for practical reasons.

Thank you for response!

**Comment by Borewit** [comment]
Interesting, I see you have quit specific needs. I would like to map these tag to metadata.common.replaygain_track_peak & metadata.common.replaygain_track_gain. Essentially to:
- read these values as number (each has a values expressed in dB & ratio)
- be independent of the input format & tagging system used

As you are an expert in this area, can you please help me with reviewing that effort?

How does that work if you broadcast with Icecast, is it streaming the MP3 as is, or does it re-encode MP3 again?

**Comment by Borewit** [comment]
`metadata.common.replaygain_track_peak` & `metadata.common.replaygain_track_gain` should now be accessible on the [branch](https://github.com/Borewit/music-metadata/tree/hack-reading-mp3-with-ape).

**Comment by jri** [comment]
> I would like to map these tag to `metadata.common.replaygain_track_peak` & `metadata.common.replaygain_track_gain` (#258). Essentially to:
> 
> * read these values as number (each has a values expressed in dB & ratio) (feature #192)
> * be independent of the input format & tagging system used
> 
> As you are an expert in this area, can you please help me with reviewing that effort?

Yes, having format & tagging system agnostic access to parsed gain values is indeed a good idea!

E.g. [aacgain](http://aacgain.altosdesign.com), the `mp3gain` pendant for AAC, creates the 4 gain tags as well, but not as APEv2 tags but as `iTunes` tags:

`tagTypes: [ 'iTunes' ]`

```
native: {
  iTunes: [
    ...
    { id: '----:com.apple.iTunes:replaygain_track_gain', value: '0.19' },
    { id: '----:com.apple.iTunes:replaygain_track_peak', value: '1.05' },
    { id: '----:com.apple.iTunes:replaygain_track_minmax', value: '67,174' },
    { id: '----:com.apple.iTunes:replaygain_undo', value: '-2,-2' }
  ]
}
```

These could be mapped to `metadata.common.replaygain_track_peak` & `metadata.common.replaygain_track_gain` as well. Also the "minmax" and "undo" values could be mapped then (for MP3/APEv2 as well). That would be quite a benefit.

As "minmax" and "undo" values are composite they could be mapped to objects, e.g.
`{min: 67, max: 174}` and `{leftChannel: -2, rightChannel: -2}`.

Format and semantics of the 4 gain tags is described here:
https://wiki.hydrogenaud.io/index.php?title=ReplayGain#MP3Gain

> How does that work if you broadcast with Icecast, is it streaming the MP3 as is, or does it re-encode MP3 again?

Re-encoding takes place indeed.
The stream relayed to the listeners has a fixed encoding and bitrate, e.g. MP3 192kbps, or AAC 128kbps. At the other hand the source material from which the stream is created can be heterogenous, that is e.g. MP3/AAC/OGG/FLAC files in all possible bitrates plus mixer/soundcard output for the live parts plus crossfades and effects applied. So re-encoding is required.

Broadcasting is split into 2 responsibilities:
1) the "source client" generates the actual stream from the source material, and transfers it to the icecast server via a TCP socket connection. The source client knows about encoders, decoders, and tagging formats, applying crossfades/effects, and how to enrich the icecast stream with "ICY" metadata to convey artist and song names. Source clients are e.g. [Liquidsoap](https://www.liquidsoap.info) and [Music Player Daemon](https://www.musicpd.org).
2) the "streaming server" (e.g. icecast) receives the stream from the source client and relays it to the connected listeners. Icecast itself knows nothing about codecs and tag formats.

**Comment by jri** [comment]
I encountered another file whose APEv2 tags are not returned, possibly because it's has VBR and an Xing header. According to `mp3val` it has tags `+ID3v1+ID3v2+APEv2, Xing header` but music-metadata only returns `tagTypes: [ 'ID3v2.3', 'ID3v1' ]`.

The file: [vbr.mp3.zip](https://github.com/Borewit/music-metadata/files/3613307/vbr.mp3.zip)

**Comment by Borewit** [comment]
Thanks a lot for your feedback @jiri, very interesting. This is the kind of feedback which helps me to improve the quality of music-metadata. I moved the mapping of the iTunes tags to #258. Related but different function and I can merge that part easily in to the release.

> Re-encoding takes place indeed.

So the best quality a listening may get is a FLAC stream, which gives the quality of the original which may be as bad as 128 kbit. Maybe not feasible, but it would be must nicer to store the recordings in FLAC, corresponding to exact studio recordings, and have the required parameters, to broadcast with stable perceived volume, stored in the metadata. In that case the only quality reduction would be the encoding corresponding to the stream a listener has chosen. Please forgive me for being naive.

> I encountered another file whose APEv2 tags are not returned

The issue is caused by an appending [Lyrics3v2](http://id3.org/Lyrics3v2) header (which is not supported yet). Xing headers are supported and are not a problem. The appending headers are the source of all evil. Those requires reading in reverse direction (reading from end of the file, backwards). With a file that is no problem, on a stream that becomes problematic.

**Comment by Borewit** [comment]
@jri I have send you a small souvenir by mail.

**Comment by jri** [comment]
> Maybe not feasible, but it would be must nicer to store the recordings in FLAC, corresponding to exact studio recordings, and have the required parameters, to broadcast with stable perceived volume, stored in the metadata.

I don't know much about FLAC, so I'm not sure if a ReplayGain implementation for FLAC exists, but I would assume so.

The scenario you describe should be feasible indeed then. I guess most ReplayGain tools support these 2 modes: 1) "Analyze Only" to analyze the file and create tags containing the resulting gain data, and 2) "Apply" which implies Analyze and additionally *applies* the calculated gain to the audio content. In case of MP3 it modifies the global gain fields in the audio data; no re-encoding required. I don't know how "Apply" would work for FLAC but in your scenario you would prefer "Analyze Only" anyways.

One advantage of "Apply" is the volume adjustment will work with every player, or, in case of broadcasting, with every source client. In contrast "Analyze Only" requires the player/source client to be ReplainGain compliant as it must read out the gain tags and apply them on-the-fly. E.g. [Music Player Daemon](https://www.musicpd.org/) *is* ReplayGain compliant.

> The issue is caused by an appending Lyrics3v2 header (which is not supported yet). Xing headers are supported and are not a problem.

That's good to know VBR and Xing headers are not a problem for music-metadata. Do you think music-metadata will support reading APEv2 tags when a lyrics header is present in the future, at least for files?

**Comment by Borewit** [comment]
If you create a new issue for the lyrics header, I will deal with it. I will at least extend the hack.
My other responsibilities are calling, so I will need a bit more time for this one.

I will consider to make parsing the "post" headers (the hack) part of the release, but the code needs to mature a bit more first.

**Comment by Borewit** [comment]
> I'm not sure if a ReplayGain implementation for FLAC

I took [foobar2000](https://www.foobar2000.org/) and [downloaded](https://www.foobar2000.org/components/view/foo_arg) the [ReplayGain component (plugin)](http://eolindel.free.fr/foobar0.9/Replaygain.php).

![image](https://user-images.githubusercontent.com/23255260/64979587-4e415f80-d8b8-11e9-8f9a-e05e22946a01.png)

Alternatively you could use [m4a/ALAC](https://wiki.hydrogenaud.io/index.php?title=ALAC), using iTunes tags. You can convert the audio between one lossless format to another (APE, WAVE, WavPack, ALAC, FLAC) without loosing any quality. The metadata, although software may do an attempt to convert as well, will suffer from it. Tags may end up as useless because may be literary converted, while format they belong to a different tag. Best is to re-write those. For this task [MusicBrainz Picard](https://picard.musicbrainz.org/) is my best friend.

Yet, FLAC is simply the most used, and best supported lossless format around. Technically very clean, clear, simple (encoding wise more simple then lossy formats). APE is also a beauty, I believe a slightly better compression ratio, but much less good supported.

MP3 is terms of serialization, documentation, extension, deviations from standards the ugliest of all formats.

**Comment by jri** [comment]
That's interesting information.
Thank you!

**Comment by jri** [comment]
I got another file whose APE tags are not returned. Can you check if it's due to a lyrics header again? (In iTunes no lyrics are shown, but that possibly doesn't matter.) `mp3val` shows `+ID3v1+ID3v2+APEv2, Xing header`; but music-metadata returns `[ 'ID3v2.3', 'ID3v1' ]`.

The file: [dream.mp3.zip](https://github.com/Borewit/music-metadata/files/3638626/dream.mp3.zip)

**Comment by Borewit** [comment]
> Can you check if it's due to a lyrics header again?

Correct, it cannot be parsed yet by the [hack](https://github.com/Borewit/music-metadata/tree/hack-reading-mp3-with-ape) due to an appending [lyrics header](https://github.com/Borewit/music-metadata/issues/264).

**Comment by Borewit** [comment]
Implemented in [v4.8.0](https://github.com/Borewit/music-metadata/releases/tag/v4.8.0)
